src/selection.ts

```diff
@@ -229,32 +183,74 @@ export interface IntervalSelectionConfig extends BaseSelectionConfig {
    * __See also:__ [`mark`](https://vega.github.io/vega-lite/docs/selection-mark.html) documentation.
    */
   mark?: BrushConfig;
+}
 
+export type SelectionTypeConfig = SingleSelectionConfig | MultiSelectionConfig | IntervalSelectionConfig;
+
+export interface BaseSelectionDef<T extends SelectionType> {
   /**
-   * Initialize the selection with a mapping between [projected channels or field names](https://vega.github.io/vega-lite/docs/project.html) and arrays of
-   * initial values.
-   *
-   * __See also:__ [`init`](https://vega.github.io/vega-lite/docs/init.html) documentation.
+   * Required. A unique name for the selection. Selection names should be valid JavaScript identifiers: they should contain only alphanumeric characters (or "$", or "_") and may not start with a digit. Reserved keywords that may not be used as parameter names are "datum", "event", "item", and "parent".
    */
-  init?: SelectionInitIntervalMapping;
-}
+  name: string;
 
-export interface BaseSelectionDef<T extends 'single' | 'multi' | 'interval'> {
   /**
    * Determines the default event processing and data query for the selection. Vega-Lite currently supports three selection types:
    *
    * - `"single"` -- to select a single discrete data value on `click`.
    * - `"multi"` -- to select multiple discrete data value; the first value is selected on `click` and additional values toggled on shift-`click`.
    * - `"interval"` -- to select a continuous range of data values on `drag`.
    */
-  type: T;
+  select:
+    | SelectionType
+    | ({
+        type: SelectionType;
+      } & (T extends 'single'
```

Comment:
Idea: maybe add more generic arguments to BaseSelectionDef to avoid this complicated type logic. Alternatively, we could have a base type that we inherit from but I'm not sure whether the scheme generator would inherit the comments (which I suppose is the goal of having the type BaseSelectionDef).

Reply:
Yeah I hear you. This logic should simplify once we collapse `single` and `multi` into a single `point` selection type. I tried to simplify further and this is where I got to – but you're savvier with TypeScript, so I'm open to suggestions :)

Reply:
I'll refactor this when it's in master. This works for now.